equistore-core/src/io/load.rs

```diff
@@ -0,0 +1,168 @@
+use std::collections::HashSet;
+use std::sync::Arc;
+
+use byteorder::{LittleEndian, BigEndian, ReadBytesExt};
+use zip::ZipArchive;
+
+use crate::{TensorMap, TensorBlock, Labels, Error, eqs_array_t};
+
+use super::check_for_extra_bytes;
+use super::labels::read_npy_labels;
+use super::npy_header::{Header, DataType};
+
+
+/// Load the serialized tensor map from the given path.
+///
+/// Arrays for the values and gradient data will be created with the given
+/// `create_array` callback, and filled by this function with the corresponding
+/// data.
+///
+/// `TensorMap` are serialized using numpy's `.npz` format, i.e. a ZIP file
+/// without compression (storage method is STORED), where each file is stored as
+/// a `.npy` array. Both the ZIP and NPY format are well documented:
+///
+/// - ZIP: <https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT>
+/// - NPY: <https://numpy.org/doc/stable/reference/generated/numpy.lib.format.html>
+///
+/// We add other restriction on top of these formats when saving/loading data.
+/// First, `Labels` instances are saved as structured array, see the `labels`
+/// module for more information. Only 32-bit integers are supported for Labels,
+/// and only 64-bit floats are supported for data (values and gradients).
+///
+/// Second, the path of the files in the archive also carry meaning. The keys of
+/// the `TensorMap` are stored in `/keys.npy`, and then different blocks are
+/// stored as
+///
+/// ```bash
+/// /  blocks / <block_id>  / samples.npy
+///                         / components  / 0.npy
+///                                       / <...>.npy
+///                                       / <n_components>.npy
+///                         / properties.npy
+///                         / values.npy
+///
+///                         # optional sections for gradients, one by parameter
+///                         /   gradients / <parameter> / samples.npy
+///                                                     /   components  / 0.npy
+///                                                                     / <...>.npy
+///                                                                     / <n_components>.npy
+///                                                     /   values.npy
+/// ```
+pub fn load<R, F>(reader: R, create_array: F) -> Result<TensorMap, Error>
+    where R: std::io::Read + std::io::Seek,
+          F: Fn(Vec<usize>) -> Result<eqs_array_t, Error>
+{
+    let mut archive = ZipArchive::new(reader).map_err(|e| ("<root>".into(), e))?;
+
+    let path = String::from("keys.npy");
+    let keys = read_npy_labels(archive.by_name(&path).map_err(|e| (path, e))?)?;
+
+    // FIXME before merging: should we add a fallback?
+    assert!(
+        archive.by_name("blocks/0/values/data.npy").is_err(),
+        "this is the old data format, TODO: fallback"
+    );
```

Comment:
I think a legacy parser is useful. We can also remove the legacy parser quite quickly again.

Reply:
I have a script to convert from old to new format using only numpy, so you don't even need equistore to convert the data. I'll push it ASAP

Reply:
And this code checks if we have a file in the old format, and point users to the script